HPXMLtoOpenStudio/resources/airflow.rb

```diff
@@ -1399,77 +1425,241 @@ def self.add_ee_for_vent_fan_power(model, obj_name, frac_lost, is_cfis, pow = 0.
+        vent_mech_sens_eff += vent_fan.average_oa_flow_rate / oa_flow_rate_sum * hrv_erv_effectiveness_map[vent_fan][:vent_mech_sens_eff]
+        vent_mech_lat_eff += vent_fan.average_oa_flow_rate / oa_flow_rate_sum * hrv_erv_effectiveness_map[vent_fan][:vent_mech_lat_eff]
+      end
+
+      # Apply ERV/HRV mechanical ventilation
+      oa_flow_rate_str = "#{UnitConversions.convert(oa_flow_rate_sum, 'cfm', 'm^3/s')}"
+      tot_flow_rate_str = "#{UnitConversions.convert(tot_flow_rate_sum, 'cfm', 'm^3/s')}"
+      program = apply_fan_load_to_infil_program(model: model, infil_program: program, oa_flow_rate_var_name: oa_flow_rate_str, tot_flow_rate_var_name: tot_flow_rate_str, vent_mech_sens_eff: vent_mech_sens_eff, vent_mech_lat_eff: vent_mech_lat_eff, fan_sens_load_actuator: fan_sens_load_actuator, fan_lat_load_actuator: fan_lat_load_actuator, preconditioned: preconditioned)
+    end
+    return program
+  end
+
+  def self.apply_preconditioning_equipment(model, vent_mech_precond, infil_program, sens_load_actuator, lat_load_actuator, cfis_fan_actuator, hrv_erv_effectiveness_map)
+    # setpoint sensors
+    htg_spt = OpenStudio::Model::EnergyManagementSystemSensor.new(model, 'Zone Thermostat Heating Setpoint Temperature')
+    htg_spt.setName("#{Constants.ObjectNameAirflow} t_htg_spt s")
+    htg_spt.setKeyName(@living_zone.name.to_s)
+
+    clg_spt = OpenStudio::Model::EnergyManagementSystemSensor.new(model, 'Zone Thermostat Cooling Setpoint Temperature')
+    clg_spt.setName("#{Constants.ObjectNameAirflow} t_clg_spt s")
+    clg_spt.setKeyName(@living_zone.name.to_s)
+
+    # Load Actuators for shared system
+    # Air property at inlet nodes on two sides
+    infil_program.addLine("Set HtgSptTemp = #{htg_spt.name}") # zone air temperature
+    infil_program.addLine("Set ClgSptTemp = #{clg_spt.name}") # zone air temperature
+
+    vent_mech_precond.each_with_index do |vent_mech, i|
+      # Create energy use actuator per system because of different fuel type can be applied
+      # FIXME: Preconditioning equipment energy end use: mech vent or hvac category? Proceed with a new end use: mech vent preconditioning
+      clg_energy_actuator = create_sens_lat_load_actuator_and_equipment(model: model, name: "shared mech vent precooling energy #{i}", space: @living_space, frac_lat: 0.0, frac_lost: 1.0, hpxml_fuel_type: vent_mech.precooling_fuel, end_use: Constants.ObjectNameMechanicalVentilationPreconditioning)
+      htg_energy_actuator = create_sens_lat_load_actuator_and_equipment(model: model, name: "shared mech vent preheating energy #{i}", space: @living_space, frac_lat: 0.0, frac_lost: 1.0, hpxml_fuel_type: vent_mech.preheating_fuel, end_use: Constants.ObjectNameMechanicalVentilationPreconditioning)
+      infil_program.addLine("Set #{clg_energy_actuator.name} = 0.0")
+      infil_program.addLine("Set #{htg_energy_actuator.name} = 0.0")
+
+      if not [HPXML::MechVentTypeERV, HPXML::MechVentTypeHRV].include? vent_mech.fan_type
```

Comment:
If cfis or erv/hrv, call its method and connect the calculations with some local variables: Q_precond_oa, Q_precond_tot, etc.